### ISSUE TITLE
Separation of components that require access to shadow

### ISSUE THREAD

**Issue by jcrawfordor** [kanidm/kanidm#3249]
I maintain a COPR build of kanidm for Fedora, and I would *like* to get this to a level of quality where I would recommend it to other people and upstream the build materials into this repo, but I've been a bit time and effort limited. Now that I've got a holiday weekend though I'm looking to clean up everything around packaging 1.4 so that I can get my machines upgraded.

Here is a current challenge: kanidm's OpenSuSE service file (which the Fedora package currently uses verbatim, but I am about to change that for a couple of reasons) adds the shadow group as a supplementarygroup to permit access to /etc/shadow. On Red Hat-like distributions, this group does not exist, and /etc/shadow is root:root and 0000. that means that uid=0 and CAP_DAC_OVERRIDE are required to read it, a very strict situation but one that's not too odd for such sensitive files and comports with the history of tools that access shadowfiles being setuid root. Of course kanidm sensibly does not want to run as root.

As a temporary solution, I may just leave the Fedora package unable to read these files, or I might experiment with setting up an ACL in a post-install script. Those are probably good enough for a COPR but a little hacky.

I think similar issues are going to come up with packaging in multiple distributions, though, and I'd like a method I feel better about. The best thing that occurs to me (and when I asked in fedora-devel Lennart seemed to agree) is separating out reading /etc/shadow into an independent binary that could run setuid. There are some other options as well but they seem more architecturally complicated.

I wanted to start a discussion on this and see if there are other ideas as well.

**Comment by yaleman** [comment]
Can't you add it to ReadOnlyPaths in the service definition? Having to have a separate binary to read a file, from a service that runs identity services, seems... a little much (We already do it for tasks, but that's WRITE capabilities)

**Comment by Firstyear** [comment]
ReadOnlyPaths is something else here @yaleman, that's related to systemd's ability to prevent you reading outside of certain locations, and preventing writes to the paths.

@jcrawfordor I did consider a separate "root process" that reads shadow and talks back to the daemon. I don't think it needs to be setuid, that's dangerous. We already have a precedent for this with the tasks daemon which *does* run as root. 

So either we extend the tasks daemon to handle the shadow reading (and in future writing), or we have something else. I think extending the tasks daemon makes the most sense.

Saying all this, this is a great example of why linux distros have terrible security oob because everything is either "you're root" or "youre not". I really wish that ACLs were the default which would resolve this whole schmozzle.

**Comment by jcrawfordor** [comment]
Yeah, for my own internal purposes, the way I want to solve this problem is by ACL. *my* machines all have ACLs enabled and working, but I'm not sure if that's universal enough (given that it's a mount option and etc.) to be something a package could rely on. A catch there is that I can't immediately think of a way to do an ACL with a systemd DynamicUser but I haven't looked into it that much, might be totally solvable.

tbh I forgot about the tasks daemon, that makes a lot of sense since it's got the special capabilities anyway.

This isn't an immediate problem for me or probably anyone, I can just leave /etc/shadow access not working on Fedora for now. But wanted to start the discussion. :smile_cat:

**Comment by Firstyear** [comment]
There is a bigger picture-plan here - we would like to have kanidm's unixd become a generic project like sssd, and to replace sssd/pam_unix by default. This is because we can the offer things like a localhost KDC which will be required for samba in the future due to the NTLM deprecation. 

So it is pretty important we get it working :)

**Comment by yaleman** [comment]
As discussed elsewhere, to clarify, you don't need `CAP_DAC_OVERRIDE`, `CAP_DAC_READ_SEARCH` would be enough, but that's full read to the entire filesystem... which is less bad but still not great 😄

**Comment by Firstyear** [comment]
It's also not a huge deal atm anyway, because if shadow can't be read it just means we don't attempt to auth the user in unixd, and we expect you still have pam_unix setup. This only matters if you want to drop pam_unixd on your system.
